packages/firebase_storage/firebase_storage/android/src/main/java/io/flutter/plugins/firebase/storage/TaskStateChannelStreamHandler.java

```diff
@@ -0,0 +1,114 @@
+  private final FirebaseStorage androidStorage;
+  private final StorageTask androidTask;
+
+  private final String TASK_STATE_NAME = "taskState";
+  private final String TASK_APP_NAME = "appName";
+  private final String TASK_SNAPSHOT = "snapshot";
+
+  private static final Executor taskExecutor = Executors.newSingleThreadExecutor();
+  private Boolean destroyed = false;
+
+  private final Object pauseSyncObject = new Object();
+  private final Object resumeSyncObject = new Object();
+  private final Object cancelSyncObject = new Object();
+
+  public TaskStateChannelStreamHandler(FirebaseStorage androidStorage, StorageTask androidTask) {
+    this.androidStorage = androidStorage;
+    this.androidTask = androidTask;
+  }
+
+  @Override
+  public void onListen(Object arguments, EventSink events) {
+    androidTask.addOnProgressListener(
+        taskSnapshot -> {
+          if (destroyed) return;
+          Map<String, Object> event = getTaskEventMap(taskSnapshot, null);
+          event.put(
+              TASK_STATE_NAME,
+              GeneratedAndroidFirebaseStorage.PigeonStorageTaskState.RUNNING.index);
+          events.success(event);
+          synchronized (resumeSyncObject) {
+            resumeSyncObject.notifyAll();
+          }
+        });
+
+    androidTask.addOnPausedListener(
+        taskSnapshot -> {
+          if (destroyed) return;
+          Map<String, Object> event = getTaskEventMap(taskSnapshot, null);
+          event.put(
+              TASK_STATE_NAME, GeneratedAndroidFirebaseStorage.PigeonStorageTaskState.PAUSED.index);
+          events.success(event);
+          synchronized (pauseSyncObject) {
+            pauseSyncObject.notifyAll();
+          }
+        });
+
+    androidTask.addOnSuccessListener(
+        taskSnapshot -> {
+          if (destroyed) return;
+          Map<String, Object> event = getTaskEventMap(taskSnapshot, null);
+          event.put(
+              TASK_STATE_NAME,
+              GeneratedAndroidFirebaseStorage.PigeonStorageTaskState.SUCCESS.index);
+          events.success(event);
+          //destroy();
+        });
+
+    androidTask.addOnCanceledListener(
+        () -> {
+          if (destroyed) return;
+          Map<String, Object> event = getTaskEventMap(null, null);
+          event.put(
+              TASK_STATE_NAME,
+              GeneratedAndroidFirebaseStorage.PigeonStorageTaskState.CANCELED.index);
+          events.success(event);
+        });
+
+    androidTask.addOnFailureListener(
+        exception -> {
+          if (destroyed) return;
+          Map<String, Object> event = getTaskEventMap(null, exception);
+          event.put(
+              TASK_STATE_NAME, GeneratedAndroidFirebaseStorage.PigeonStorageTaskState.ERROR.index);
+          events.error(
+              FlutterFirebaseStoragePlugin.DEFAULT_ERROR_CODE, exception.getMessage(), event);
+        });
+  }
+
+  @Override
+  public void onCancel(Object arguments) {
```

Comment:
When a task is completed/cancelled/failed. This needs to be called from the `FlutterFirebaseStorageTask`, also need to clean up from the Maps holding them [here](https://github.com/firebase/flutterfire/pull/11521/files#diff-9de9c3a74a36c16b77f4d83c24fe808819acbb88a6ff4f6f59a5e882da49c438R47-R48)

Reply:
I would put the clean up of a stream handler in the destroy method here: https://github.com/firebase/flutterfire/pull/11521/files#diff-b6e5a6bea234c57c8a9a1bf39f759509722c844490626f89f7f7575f2ecf8169R140

Reply:
done

Reply:
The `onCancel()` method needs to be called whenever the task is completed/cancelled/failed. We also need to remove it from the `eventChannels` map and `streamHandlers`.

Making a function like below:
```java
private void removeEventListener(String identifier) {
      eventChannels.get(identifier).setStreamHandler(null);
      streamHandlers.get(identifier).onCancel(null);
  }
```
and call it from `destroy()` method in the `FlutterFirebaseStorageTask` class. Otherwise, the only time the stream handlers are removed are when the Flutter engine is detached.

You can also remove [this `destroy()`](https://github.com/firebase/flutterfire/blob/870432c6336838209c63965ba2107699343e95c0/packages/firebase_storage/firebase_storage/android/src/main/java/io/flutter/plugins/firebase/storage/TaskStateChannelStreamHandler.java#L94) method because the task is already destroyed in `onDetachfromEngine`, `didReinitializeFirebaseCore` & when we call `destroy()` in the `addOnFailureListener`, `addOnCanceledListener` & `addOnSuccessListener` listeners which are the times it needs to be destroyed. The issue might be that we call `flutterTask.destroy()` on a task that has already been removed from the `inProgressTasks` static array.

In fact, it might be worth adding this line to the top of the `destroy()` method:

```java
 void destroy() {
// The line below:
    if(destroyed) return;
    destroyed = true;

    synchronized (inProgressTasks) {
      if (storageTask.isInProgress() || storageTask.isPaused()) {
        storageTask.cancel();
      }
      inProgressTasks.remove(handle);
    }

    synchronized (cancelSyncObject) {
      cancelSyncObject.notifyAll();
    }

    synchronized (pauseSyncObject) {
      pauseSyncObject.notifyAll();
    }

    synchronized (resumeSyncObject) {
      resumeSyncObject.notifyAll();
    }
  }

```